### PR TITLE
Filter out unneeded labels

### LINKED ISSUE #605: Propagated labels causes issues for Deployment, StatefulSet and DaemonSet
Some labels like the ones used by GitOps tools and other controllers that would possible create OTEL resources such as and not limited to Flux V2, ArgoCD, and others will create issues for immutable labels passed to `X.spec.template.metadata.labels` giving errors like below

```
{
    "level": "error",
    "ts": 1638708207.8387449,
    "logger": "controller-runtime.manager.controller.opentelemetrycollector",
    "msg": "Reconciler error",
    "reconciler group": "opentelemetry.io",
    "reconciler kind": "OpenTelemetryCollector",
    "name": "main",
    "namespace": "opentelemetry-system",
    "error": "failed to reconcile the expected daemon sets: failed to apply changes: DaemonSet.apps \"main-collector\" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{\"app.kubernetes.io/component\":\"opentelemetry-collector\", \"app.kubernetes.io/instance\":\"opentelemetry-system.main\", \"app.kubernetes.io/managed-by\":\"opentelemetry-operator\", \"app.kubernetes.io/name\":\"main-collector\", \"app.kubernetes.io/part-of\":\"opentelemetry\", \"kustomize.toolkit.fluxcd.io/name\":\"opentelemetry-system-codegen\", \"kustomize.toolkit.fluxcd.io/namespace\":\"flux-system\"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable",
    "stacktrace": "sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:214"
}
```

### PR DESCRIPTION
**Pull request by ekarlso**
Fixes #605

### CODE REVIEW

File: pkg/collector/labels.go [hunk 1]
```diff
@@ -16,17 +16,29 @@ package collector
 
 import (
 	"fmt"
+	"regexp"
 
 	"github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1"
 )
 
+func isFilteredLabel(label string, filterLabels []string) bool {
```
Comment (jpkrohling) [inline pkg/collector/labels.go:24]: This seems well-tested via the collector/daemonset test suites, but I think it does deserve a new table-test as part of the label's test suite.

File: pkg/collector/labels_test.go [hunk 2]
```diff
@@ -50,9 +50,26 @@ func TestLabelsPropagateDown(t *testing.T) {
...
+	otelcol := v1alpha1.OpenTelemetryCollector{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{"test.bar.io": "foo", "test.foo.io": "bar"},
+		},
+	}
+
+	// This requires the filter to be in regex match form and not the other simpler wildcard one.
+	labels := Labels(otelcol, []string{".*.bar.io"})
+
+	t.Log(labels)
```
Comment (jpkrohling) [inline pkg/collector/labels_test.go:70]: t.Log? Is this perhaps a leftover from a local debug?

File: main.go [hunk 3]
```diff
@@ -91,11 +93,14 @@ func main() {
...
 	pflag.StringVar(&autoInstrumentationNodeJS, "auto-instrumentation-nodejs-image", fmt.Sprintf("ghcr.io/open-telemetry/opentelemetry-operator/autoinstrumentation-nodejs:%s", v.AutoInstrumentationNodeJS), "The default OpenTelemetry NodeJS instrumentation image. This image is used when no image is specified in the CustomResource.")
 	pflag.StringVar(&autoInstrumentationPython, "auto-instrumentation-python-image", fmt.Sprintf("ghcr.io/open-telemetry/opentelemetry-operator/autoinstrumentation-python:%s", v.AutoInstrumentationPython), "The default OpenTelemetry Python instrumentation image. This image is used when no image is specified in the CustomResource.")
+	pflag.StringArrayVar(&labelsFilter, "labels", []string{}, "Labels to filter away from propagating onto deploys")
 	pflag.Parse()
 
 	logger := zap.New(zap.UseFlagOptions(&opts))
 	ctrl.SetLogger(logger)
 
+	pflag.Parse()
```
Comment (jpkrohling) [inline main.go:102]: Why is it parsing the flags once again?

**Review by jpkrohling — APPROVED** [review]
LGTM, but shouldn't this be documented somewhere?

**Comment by ekarlso** [comment]
> LGTM, but shouldn't this be documented somewhere?

I think it will totally depend on you setup but yeah if you want me to add docs please say where :)

**Pull request merged**